File: python_patterns/code/SRect.py
try:
    reload (SPoint) 
except NameError:
    import SPoint 

#==============================================================================
# The SRect class
#==============================================================================
class SRect:
    def __init__(self):
        self.left   = 0
        self.right  = 0
        self.top    = 0
        self.bottom = 0
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~            
    def __str__(self):
        return "Rect: (%f,%f) to (%f,%f)" % (self.left, self.bottom, self.right, self.top)
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~            
    def fromPoint(self,pt):
        """
        Initialise from a SPoint structure
        """
        self.left   = pt.x
        self.right  = pt.x
        self.top    = pt.y
        self.bottom = pt.y
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~            
    def includePoint(self,pt):
        """
        Expand to include a SPoint structure
        """
        self.left   = min (pt.x, self.left)
        self.right  = max (pt.x, self.right)
        self.top    = max (pt.y, self.top)
        self.bottom = min (pt.y, self.bottom)
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~            
    def includeRect(self,rect):
        """
        Expand to include a SRect structure
        """
        self.left   = min (rect.left, self.left)
        self.right  = max (rect.right, self.right)
        self.top    = max (rect.top, self.top)
        self.bottom = min (rect.bottom, self.bottom)
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~            
    def display(self):
        print self
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~            
    def width(self):
        return self.right - self.left
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~            
    def height(self):
        return self.top - self.bottom
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~            
    def centre(self):
        xc = float (self.right + self.left) / 2
        yc = float (self.top + self.bottom) / 2
        return SPoint.SPoint(xc, yc)
    
    

        
        

File: svg/colouringin/srect.js
//==============================================================================
// The SRect class
//==============================================================================
SRect = function ()
{
    this.left   = 0;
    this.right  = 0;
    this.top    = 0;
    this.bottom = 0;
}
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
SRect.prototype.toString = function ()
{
    return "Rect: (" + this.left + "," + this.bottom + ") to (" + this.right + "," + this.top + ")";
}
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Initialise from a SPoint structure
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
SRect.prototype.fromPoint = function (pt)
{
    this.left   = pt.x;
    this.right  = pt.x;
    this.top    = pt.y;
    this.bottom = pt.y;
}
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Expand to include a SPoint
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
SRect.prototype.includePoint = function (pt)
{
    this.left   = Math.min (pt.x, this.left);
    this.right  = Math.max (pt.x, this.right);
    this.top    = Math.max (pt.y, this.top);
    this.bottom = Math.min (pt.y, this.bottom);
}
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Expand to include a SRect structure
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
SRect.prototype.includeRect = function (rect)
{
    this.left   = Math.min (rect.left, this.left);
    this.right  = Math.max (rect.right, this.right);
    this.top    = Math.max (rect.top, this.top);
    this.bottom = Math.min (rect.bottom, this.bottom);
}
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
SRect.prototype.display = function ()
{
    return this.toString ();
}
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
SRect.prototype.width = function ()
{
    return this.right - this.left;
}
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
SRect.prototype.height = function ()
{
    return this.top - this.bottom;
}
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
SRect.prototype.centre = function ()
{
    var xc = (this.right + this.left) / 2;
    var yc = (this.top + this.bottom) / 2;
    return new SPoint(xc, yc);
}



